lib/common/ipc_client.c

```diff
@@ -1361,44 +1376,18 @@ crm_ipc_send(crm_ipc_t *client, const xmlNode *message,
         goto send_cleanup;
     }
 
-    /* Read the IPC response, respecting any timeout we were passed */
-    if (ms_timeout > 0) {
-        rc = internal_ipc_get_reply(client, header->qb.id, ms_timeout, &bytes);
-        if (rc != pcmk_rc_ok) {
-            /* We didn't get the reply in time, so disable future sends for now.
-             * The only alternative would be to close the connection since we
-             * don't know how to detect and discard out-of-sequence replies.
-             *
-             * @TODO Implement out-of-sequence detection
-             */
-            client->need_reply = TRUE;
-        }
-        rc = (int) bytes; // Negative system errno, or size of reply received
-
-    } else {
-        do {
-            qb_rc = qb_ipcc_recv(client->ipc, client->buffer, client->buf_size,
-                                 -1);
-        } while ((qb_rc == -EAGAIN) && crm_ipc_connected(client));
-        rc = (int) qb_rc; // Negative system errno, or size of reply received
+    rc = internal_ipc_get_reply(client, header->qb.id, ms_timeout, &bytes, reply);
+    if (rc != pcmk_rc_ok) {
+        /* We didn't get the reply in time, so disable future sends for now.
+         * The only alternative would be to close the connection since we
+         * don't know how to detect and discard out-of-sequence replies.
+         *
+         * @TODO Implement out-of-sequence detection
+         */
+        client->need_reply = TRUE;
     }
 
-    /* Process the response, or error if we couldn't read it */
-    if (rc > 0) {
-        pcmk__ipc_header_t *hdr = (pcmk__ipc_header_t *)(void*)client->buffer;
-
-        crm_trace("Received %d-byte reply %d to %s IPC %d: %.100s",
-                  rc, hdr->qb.id, client->server_name, header->qb.id,
-                  crm_ipc_buffer(client));
-
-        if (reply) {
-            *reply = pcmk__xml_parse(crm_ipc_buffer(client));
-        }
-
-    } else {
-        crm_trace("No reply to %s IPC %d: rc=%d",
-                  client->server_name, header->qb.id, rc);
-    }
+    rc = (int) bytes; // Negative system errno, or size of reply received
```

Comment:
Before this point, `rc` is set to `pcmk_rc_ok` on success (`*bytes > 0`) or when `*bytes == 0` (is this success or not?), or to the positive system errno on failure (when `*bytes < 0`)

Reply:
Fixed, I think.  Error handling in these functions is more annoying than I would like.